### ISSUE TITLE
如何使用WeightedRandomSampler和DistributedBatchSampler创建分布式数据采样器

### ISSUE THREAD

**Issue by xywlpo** [PaddlePaddle/Paddle#42870]
### 请提出你的问题 Please ask your question

您好，请问如何使用DistributedBatchSampler构建一个WeightedRandomSampler的分布式数据采集器，将其送入DataLoader创建数据加载器

**Comment by WenmuZhou** [comment]
你好，这样写
```python
import numpy as np
from paddle.io import Dataset, DataLoader, DistributedBatchSampler, WeightedRandomSampler

# init with dataset
class RandomDataset(Dataset):
    def __init__(self, num_samples):
        self.num_samples = num_samples

    def __getitem__(self, idx):
        image = np.random.random([784]).astype('float32')
        label = np.random.randint(0, 9, (1, )).astype('int64')
        return image, label

    def __len__(self):
        return self.num_samples

dataset = RandomDataset(5)
sampler = WeightedRandomSampler(weights=np.array([0.1, 0.3, 0.5, 0.7, 0.2]),
                                num_samples=5,
                                replacement=True)
batch_sampler = DistributedBatchSampler(sampler, batch_size=2)

data_loader = DataLoader(
        dataset=dataset,
        batch_sampler=batch_sampler
       )

for i, (image, label) in enumerate(data_loader()):
      print(label)
```

**Comment by xywlpo** [comment]
> 你好，这样写
> 
> ```python
> import numpy as np
> from paddle.io import Dataset, DataLoader, DistributedBatchSampler, WeightedRandomSampler
> 
> # init with dataset
> class RandomDataset(Dataset):
>     def __init__(self, num_samples):
>         self.num_samples = num_samples
> 
>     def __getitem__(self, idx):
>         image = np.random.random([784]).astype('float32')
>         label = np.random.randint(0, 9, (1, )).astype('int64')
>         return image, label
> 
>     def __len__(self):
>         return self.num_samples
> 
> dataset = RandomDataset(5)
> sampler = WeightedRandomSampler(weights=np.array([0.1, 0.3, 0.5, 0.7, 0.2]),
>                                 num_samples=5,
>                                 replacement=True)
> batch_sampler = DistributedBatchSampler(sampler, batch_size=2)
> 
> data_loader = DataLoader(
>         dataset=dataset,
>         batch_sampler=batch_sampler
>        )
> 
> for i, (image, label) in enumerate(data_loader()):
>       print(label)
> ```

感谢您的回复，我看paddle的文档上DistributedBatchSampler的接口，要求第一个输入的是一个dataset (paddle.io.Dataset) ，可以直接传入采样器吗？

**Comment by WenmuZhou** [comment]
你好，我本地测试了一下，上面的写法虽然能跑，但是WeightedRandomSampler没起到作用，目前还不支持他们两一起用，抱歉，我们会尽快支持他们两一起用

**Comment by xywlpo** [comment]
> 你好，我本地测试了一下，上面的写法虽然能跑，但是WeightedRandomSampler没起到作用，目前还不支持他们两一起用，抱歉，我们会尽快支持他们两一起用

请问那是否有其他的写法，可以实现WeightedRandomSampler的分布式使用，感谢

**Comment by WenmuZhou** [comment]
暂时没有哈

**Comment by xywlpo** [comment]
> 暂时没有哈

您好，我参考pytorch的方式写了一下，但是还有些问题，期望您的帮助，我改写的如下：
<img width="529" alt="1" src="https://user-images.githubusercontent.com/728405/171015164-cf7e462a-88c1-420e-8ee3-69a06213ff82.png">
<img width="741" alt="2" src="https://user-images.githubusercontent.com/728405/171015176-0dcbbc06-001b-444b-a7db-1b81bdd0c530.png">

def main():
  ……
  weights = np.array([w_neg if label == 1 else w_pos for data, label in self.dataset])
  weights_sampler = WeightedRandomSampler(weights, total_nums)
  self._batch_sampler = DistributedSamplerWrapper(weights_sampler, self.batch_size)
  self.dataloader = DataLoader(dataset=self.dataset,
                                    batch_sampler=self._batch_sampler,
                                    collate_fn=_batch_transforms,
                                    num_workers=self.worker_num,
                                    return_list=return_list,
                                    use_shared_memory=use_shared_memory)
  self.loader = iter(self.dataloader)
  return self

然后在最后返回时报错如下。
<img width="737" alt="WX20220530-223137@2x" src="https://user-images.githubusercontent.com/728405/171014325-6c52c906-76c0-4bb1-8f8f-70a0f99be4fa.png">
这个方式在pytorch上是ok的，但是这里就会报错，请问是什么原因？

**Comment by WenmuZhou** [comment]
你好，目前不能这么实现哈，我们这边后续会添加官方实现，到时候会在此回复

**Comment by CSberlin** [comment]
> 

你好, 请问实现了嘛?
